### DISCUSSION TITLE
getStaticProps request context (host) when on Vercel

### DISCUSSION THREAD

**Discussion by adriano-tirloni** [vercel/next.js#12848] [Help] [answered]
My app uses the whole url, including subdomain and domain to request the correct data / styling data from the backend. 

When using `getStaticProps` + `getStaticPaths with fallback: true` in production with Vercel is it possible to know the `host` from the request that executed the serverless funciton? 

Much like `getServerSideProps`, which I was able to make it work correctly.
Imagine this:

redsite.wiki.com/what-is-a-byte
bluesite.wiki.com/what-is-a-byte

**Comment by leerob** [comment]
`getStaticProps` runs at **build time** (unless using [Incremental Static Regeneration](https://nextjs.org/blog/next-9-4#incremental-static-regeneration-beta)) whereas `getServerSideProps` runs at **request time**. 

If you need to fetch data on the client-side for a static page, you might consider serving a static shell (loading state) and using [SWR](https://swr.now.sh/) to fetch data. I'd need to hear more about what you're trying to build 😄

**Reply by adriano-tirloni** [reply]
Hey @leerob, the use case would be Incremental Static Regeneration, like the example they provided (https://github.com/lfades/static-tweet) using `getStaticProps + getStaticPaths with fallback: true` to generate the static page at the first request, and regenerate if content changed (which for an article or CMS it rarely does)

The only difference is that all the information they need for the example comes via props, but I use the hostname and the path to render the page.

I do pure ServerSideRendering today, but the content is 99% static, and Incremental would bring a lot of benefits.

**Reply by leerob** [reply]
Gotcha! That makes more sense. 

Based on [the docs](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation), it doesn't seem like you have access to `req`. 

Where are you pulling data from in `getStaticProps`?

**Reply by eMKoded** [reply]
Any progress in this issue? I have exactly the same case as @adriano-tirloni. I managed to rewrite urls from /sample/[id] to /[host]/sample/[id] but it only works in the build time. When new previously unknown [id] comes I end up on fallback version of the page and JS error: 

`The provided as value (/sample/3311) is incompatible with the href value (/[host]/sample/[id])`

Any idea how to overcome this? SSR is last resort...

**Reply by amfeng** [reply]
@eMKoded - I'm curious if you ever ended up figuring this out? I'm also running into the incompatible href issue, sadly. How did you rewrite your urls from /sample/[id] to /[host]/sample/[id]?

**Reply by eMKoded** [reply]
You can use `customServer` where you have access to `req` and `res` so you can prepare params to next's handler as you wish. Finally I ended up with SSR and custom cache.

**Comment by iNishant** [comment]
I have a similar use case. Using the subdomain from host, would want to fetch some app constants. Use case is similar to how slack workspaces work. Different workspaces on different subdomains.

**Reply by rwproj** [reply]
Same story. Thinking of trying to use async-hooks for this to get domain prefix from async hook context on server

**Reply by iNishant** [reply]
@rwproj using `getInitialProps` or `getServerSideProps`, this is pretty simple to do because you have access to `req` object and hence you can get the subdomain from the host(`req.headers.host`). Problem with this approach is re-render on each request even though the content, in my use case, is mostly static == changes infrequently.

Can't use "serving a static shell (loading state) and using SWR to fetch data." as suggested by @leerob as my use case is for landing pages hence SEO will be impacted.

**Reply by tomsoderlund** [reply]
@rwproj Did you get it to work with `getStaticProps` and async-hooks?

**Comment by ramiel** [comment]
This is very old, is there any news on this? is there in the road map any idea to support this use case?

**Reply by steven-tey** [reply]
Hey @ramiel! As mentioned below, we recently launched the Platforms Starter Kit that's should solve this use case – it's a comprehensive template for building multi-tenant applications with built-in custom domains support (+ unlimited custom domains 😉): https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Comment by threehams** [comment]
If you don't need both subdomain SSG _and_ i18n at the same time, you can repurpose that system.

```
module.exports = {
  i18n: {
    locales: ["www", "partner"],
    domains: [
      {
        domain: "partner.domain.com",
        defaultLocale: "partner",
      },
      {
        domain: "www.domain.com",
        defaultLocale: "www",
      },
    ],
    defaultLocale: "www",
    localeDetection: false,
  },
};
```

This is absolutely not intended use, and using this will prevent you from doing localization, but the concepts are the same - segment static pages based on a property derived from the host.

**Reply by petrbela** [reply]
Isn't the locale system static (all the possible subdomains need to be known at compile time)? That would not work for dynamic subdomains that users can configure (as a "slug" in the database, for example).

**Reply by threehams** [reply]
Correct. This is not an ideal solution - more of a workaround for some cases since the alternatives (custom server, disabling SSG entirely) are far worse.

**Comment by wmattei** [comment]
I think this discussion should be answered, as it seems like a next.js limitation:
https://github.com/vercel/next.js/discussions/21431

**Reply by steven-tey** [reply]
Hey @wmattei! As mentioned below, we recently launched the Platforms Starter Kit that's should solve this use case – it's a comprehensive template for building multi-tenant applications with built-in custom domains support (+ unlimited custom domains 😉): https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Comment by surjithctly** [comment]
I hope this solves the problem: https://multi-domain-first.vercel.app/ by leerob

**Reply by iNishant** [reply]
@surjithctly https://multi-domain-first.vercel.app/ by @leerob is a pretty cool idea but this isn't ideal for pages like landing pages (or other marketing pages) if I understand `stale-while-revalidate` correctly. Because the first request for any user is still a pure SSR call and that's where its crucial to load the page fast. It works like static after that (till swr period) as explained by @leerob.

Obviously, we can't get request context at build time so I feel what we actually need is an advanced version of `getStaticPaths` where you can not only iterate over some dynamic paths at build time but also iterate over some dynamic domains.

@leerob @timneutkens what you do think about this idea?

**Reply by Skarian** [reply]
Any update on this? @leerob @timneutkens

**Reply by steven-tey** [reply]
Hey @Skarian! As mentioned below, we recently launched the Platforms Starter Kit that's should solve this use case – it's a comprehensive template for building multi-tenant applications with built-in custom domains support (+ unlimited custom domains 😉): https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Comment by tomsoderlund** [comment]
@adriano-tirloni Here’s an example of hosting **multiple domains on the same Next.js site** (while maintaining **multiple languages** and static site generation/SSG), using Next.js’ i18n system:

https://github.com/tomsoderlund/nextjs-multi-domain-locale

**Comment by acorcutt** [comment]
You can use the host in rewrites in 10.2 https://nextjs.org/docs/api-reference/next.config.js/rewrites#header-cookie-and-query-matching so just send it through as a page param and build all pages under `pages/[host]/`

```
async rewrites() {
    return [
      {
        has: [
          {
            type: 'host',
            value: '(?<host>.*)',
          },
        ],
        source: '/',
        destination: '/:host',
      },
      {
        has: [
          {
            type: 'host',
            value: '(?<host>.*)',
          },
        ],
        source: '/:path*',
        destination: '/:host/:path*',
      },
    ];
  },
```

Then its available in static props:
```
export async function getStaticProps(context) {
  return {
    props: {
      host: context.params.host,
    },
  };
}
```

With fallbacks or blocking:
```
export async function getStaticPaths() {
  return {
    paths: [],
    fallback: 'blocking',
  };
}
```
Example: https://github.com/acorcutt/next-multi-host

**Reply by nyn0x** [reply]
I got `Error: A required parameter (host) was not provided as a string in getStaticPaths for /[host]/[[...slug]]`  🤔 I try test.localhost:3000/test

**Reply by acorcutt** [reply]
Give this a try:
https://github.com/acorcutt/next-multi-host
Multiple host test:
https://next-multi-host.vercel.app/
https://next-multi-host-acorcutt.vercel.app/posts/a

Webpack 5 was giving me issues so using 4, and remove index.js from /pages/ root.

Will add some async data to test the caching & fallbacks.

**Reply by tomsoderlund** [reply]
@acorcutt Awesome approach!

**Reply by acorcutt** [reply]
I've made a couple of changes to add a static built path and caching which seems to work.

I think using a single `:slug*` to `[[...slug]]` path or mixing with others is causing issues, I've moved the catch all path into its own route. It looks like you need to rewrite every page route you want to match to keep all the other parameters consistent with the routes.

**Reply by ghost** [reply]
@acorcutt Great work!

Regarding the last issue with :slug*, do you know any way to still implement a catch-all path without /slugs/?

**Reply by acorcutt** [reply]
@Warqus The catch all seems to work if there are no other page routes below it...

A single `':host/:slug*'` to just `'/[host]/[[..slug]].js'` on the root worked ok for me, but trying to use the same `':host/:slug*'` catch all for say `'/[host]/posts/[id].js'` did not.

Try the first commit I did this is just a single catch all: https://github.com/acorcutt/next-multi-host/tree/e675a3edb606d1404d7b4d43102bfb8e6f4609de

**Reply by ghost** [reply]
@acorcutt Thanks for the reply. Unfortunately I need both. 

Is this intentionally done this way? I can imagine that if a source is recognized, it should not be rewritten again. So if we would put

```
{
    has: [
      {
        type: 'host',
        value: '(?<host>.*)',
      },
    ],
    source: '/:path*',
    destination: '/:host/:path*',
}
```

at the end of the rewrite array, all the other routes are checked first. Then whenever no route is recognized to rewrite, the catch all statement is used (and only once).

**Reply by acorcutt** [reply]
@Warqus Yes I thought it would work like that but order did not seem to matter.. the catch all route to [host]/post/[id] would load in dev but the params were messed up and the builds failed. 

Maybe try splitting the catch all and know routes to a different destination folder...
```
{
    has: [
      {
        type: 'host',
        value: '(?<host>.*)',
      },
    ],
    source: '/posts/:id',
    destination: '/knownroutes/:host/posts/:id',
},
{
    has: [
      {
        type: 'host',
        value: '(?<host>.*)',
      },
    ],
    source: '/:path*',
    destination: '/catchallroutes/:host/:path*',
}
```

**Reply by acorcutt** [reply]
@Warqus I just tried that & it still breaks https://next-multi-host-pno209z3a-acorcutt.vercel.app/

If the catch all is not too deep you could try [a]/[b]/[c]/[d] rather than a wildcard.

**Reply by ghost** [reply]
@acorcutt I will try that tomorrow! Thanks for the help. Hopefully it will work :)

**Reply by acorcutt** [reply]
@Warqus It looks like the problem is the dynamic rewrites are running before the others on the server - client side routing works replacing the wildcard with [a]/[b]/[c]/[d] style routes: https://next-multi-host-pdicl46ht-acorcutt.vercel.app/ https://github.com/acorcutt/next-multi-host/tree/catchall

**Reply by ghost** [reply]
@acorcutt I have been trying for quite a while today, but I don't get it working! Thanks a lot for your input, I will continue tomorrow. Once it works, I will let it know!

**Reply by ghost** [reply]
@acorcutt This is an example of context.params which I 'console.log' in Vercel in getStaticProps:

![image](https://user-images.githubusercontent.com/14151318/116827600-22144880-ab9a-11eb-9525-ce00547e32ff.png)

with:

![image](https://user-images.githubusercontent.com/14151318/116827611-2f313780-ab9a-11eb-8c01-b376b010eff4.png)

with beforeFiles done like this:

```
beforeFiles: [
    {
        source: "/sitemap.xml",
        destination: "/api/sitemap",
    },
    {
        source: "/sitemap/:slug*.xml",
        destination: "/api/sitemap/:slug*",
    },

    // index
    {
        has: [
            {
                type: 'host',
                value: '(?<host>.*)',
            },
        ],
        source: '/',
        destination: '/hosts/:host',
    },
]
```

with afterFiles done like this:

```
afterFiles: [
  {
      has: [
          {
              type: 'host',
              value: '(?<host>.*)',
          },
      ],
      source: '/:lang',
      destination: '/hosts/:host/catch_all/:lang/__/__/__/__',
  },
  {
      has: [
          {
              type: 'host',
              value: '(?<host>.*)',
          },
      ],
      source: '/:lang/:slug1',
      destination: '/hosts/:host/catch_all/:lang/:slug1/__/__/__',
  },
  {
      has: [
          {
              type: 'host',
              value: '(?<host>.*)',
          },
      ],
      source: '/:lang/:slug1/:slug2',
      destination: '/hosts/:host/catch_all/:lang/:slug1/:slug2/__/__',
  },
  {
      has: [
          {
              type: 'host',
              value: '(?<host>.*)',
          },
      ],
      source: '/:lang/:slug1/:slug2/:slug3',
      destination: '/hosts/:host/catch_all/:lang/:slug1/:slug2/:slug3/__',
  },
  {
      has: [
          {
              type: 'host',
              value: '(?<host>.*)',
          },
      ],
      source: '/:lang/:slug1/:slug2/:slug3/:slug4',
      destination: '/hosts/:host/catch_all/:lang/:slug1/:slug2/:slug3/:slug4',
  },
]
```

But it should be for example:

```
params = {
  host: 'test1.example.com',
  lang: 'en',
  slug1: '__',
  slug2: '__',
  slug3: '__',
  slug4: '__',
}
```

**Reply by acorcutt** [reply]
@Warqus I think this is a problem with the deploy to Vercel. I've sent a support request. It works locally as expected with next or vercel dev`. Try https://next-multi-host-pdicl46ht-acorcutt.vercel.app/ with js on/off - it looks like when deployed the server renders the wildcard routes out of order with the client and how it's running locally so SSR & client are trying to render different routes. Setting rewrites in vercel.json has same problem.

**Reply by ghost** [reply]
Your website is functioning properly in my browser with js on. Whenever I go to a random path, it renders.. I have implemented the same but it doesn’t. Any idea why that is the case? Did you do anything special when deploying?

**Reply by lucas-barros** [reply]
@acorcutt I'm trying to use another header in the rewrites, but seems only the host header is available. Have you experienced this?

**Reply by acorcutt** [reply]
@lucas-barros I've added an example with headers - this worked fine local but not deployed I'm getting `$cache` in place for the real header. Also, some headers broke the page match completely like the user-agent example. One possibility would be to use a catch all and combine them into [[...headers]] but that gives us the problem we're having above with wildcards.

@Warqus My example is still wrong, its just some combinations throw an error others fail silently and mess up the params order.

**Reply by acorcutt** [reply]
@Warqus A temporary solution might be to put your catch-all route in the fallback section and proxy it to another app (or another domain for itself?) this works for me:

```
      fallback: [
        {
          has: [
            {
              type: 'host',
              value: '(?<host>.*)',
            },
          ],
          source: '/:slug*',
          destination: 'https://next-multi-host.vercel.app/hosts/:host/slugs/:slug*',
        },
      ],
```

**Reply by ghost** [reply]
@acorcutt Thanks for the suggestion! 

I have decided to postpone a catch-all for ISR until things are fixed at Vercel. I am currently working on making my whole website ISR with the rewrite method and I'll keep the catch-all route SSR with cache for now.

**Reply by wzulfikar** [reply]
Hey @Warqus, thanks for posting. Glad I'm not crazy for thinking why my setup works using `next start` (or `next dev`) but not when deployed to vercel. FWIW, it turns out that the setup doesn't work too with `vc dev`. I end up deploying my site using docker just to see if it actually works (it works!). 

Is there any Github issue for this so I can follow the updates?

**Reply by acorcutt** [reply]
@lucas-barros I've updated the example to include headers & query string but its still not working correctly, firstly you need rewrites for every combination of header you use existing & not existing or they don't match - so the page routes will get quite complex, and also they are not available when it does client side navigation so you need to handle that case.

@wzulfikar I think the problem is related to https://github.com/vercel/next.js/blob/master/errors/rewrite-auto-export-fallback.md it might be best to stick to `getServerSideProps` for anywhere wildcards overlap with dynamic routes.

**Reply by lucas-barros** [reply]
@acorcutt ok, thank you for the information, I'll take a look.

**Reply by wzulfikar** [reply]
@acorcutt thanks for the pointer!

**Reply by styxlab** [reply]
@acorcutt that's amazing! Will try your approach!

**Reply by tomsoderlund** [reply]
@acorcutt Have you tried your approach with multiple locales?

**Reply by acorcutt** [reply]
@tomsoderlund Not yet I will give it a try with the built in locales, but you can access the accept-language header directly so a `/:host/:lang/...` route would be possible.

**Reply by zakariamofaddel** [reply]
Even tho this is a smart way of getting around the problem, it's still just a "hacky" way to get it, not how it's supposed to naturally work, since It's not a solution for big applications already built. Moving every page to be under a [host] subfolder is not doable :(

**Reply by styxlab** [reply]
@zakariamofaddel : I agree that a new `host` prop in `getStaticProps()` would be preferable, but I am really happy about the workaround solution. Moving your routes into a new [host] subfolder is really not that difficult, I think you can do this for both small and big applications.

A more complicated but more flexible solution is to make a serverless proxy endpoint with `createProxyMiddleware` from the `http-proxy-middleware` package. You can use the proxy for arbitrary rewrite and header changes. Note that this solution counts agianst your serverless execution time.

**Reply by thadeucity** [reply]
It is working flawlessly for Webpack 4, 🚀 

But I'm am curious does anyone know why this approach does not work with Webpack5?

**Reply by styxlab** [reply]
> It is working flawlessly for Webpack 4, 🚀 
> 
> 
> 
> But I'm am curious does anyone know why this approach does not work with Webpack5?

It works for me with webpack 5 and nextjs 11. Can you show your config?

**Reply by acorcutt** [reply]
@thadeucity I've updated the example to nextjs 11 & webpack 5 its working ok for me now - I think the issue was when switching between webpack 4 & 5 in dev, try removing your .next build folder and doing a clean build.

**Reply by thadeucity** [reply]
Removing the node_modules and .next folders and building it again worked for me. Thank you guys!!!

**Reply by rambabusaravanan** [reply]
@acorcutt It's working like a charm for me. Really the simplest and awesome approach dude. Thanks a lot
I am using next 11.0.0 and webpack 5

**Reply by steven-tey** [reply]
Hey guys! As mentioned below, we recently launched the Platforms Starter Kit that's should solve this use case – it's a comprehensive template for building multi-tenant applications with built-in custom domains support (+ unlimited custom domains 😉): https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Reply by acorcutt** [reply]
@steven-tey Hi - does this mean the 50 domain limit is gone or is this different from the per-project domains on the [pricing](https://vercel.com/pricing) page?

**Reply by leerob** [reply]
Hey @acorcutt - happy to share there's now unlimited custom domains on Pro. 

https://vercel.com/changelog/unlimited-custom-domains-for-all-pro-teams

**Comment by threehams** [comment]
As of Next 11, `beforeFIles` doesn't stop at the first rewrite - it loops through all of them in a cascading way.

I'm currently using `fallback` as a workaround, but that means there's a chance for leads to unexpected matches:

```
pages/
  [host]/
    user/
      account/
        claim.tsx
    account/
      [section].tsx
```

next.config.js
```
{
  fallback: [
    {
      source: "/:path*",
      destination: "/:host/:path*",
      has: [
        {
          type: "host",
          value: "(?<host>.*)",
        },
      ],
    },
  ]
}
```

Visiting `/user/account/claim/` is meant to hit that rewrite and land on `/[host=:host]/user/account/claim`, but it ends up at `/[host=user]/account/[section=claim]`.

If you're using `fallback` like this and hitting similar problems, I think it's possible to now move this to the end of `beforeFiles` instead.

**Reply by acorcutt** [reply]
@threehams Yes - if possible avoid using `/:wildcard*` matches that overlap with multiple routes, it does mean you have to add rewrites that match for every single route but will be safer.

**Comment by styxlab** [comment]
As of Next 12, this can now be accomplished in the edge handler `_middleware.ts`. This is a lot more transparent. Have a look at the [hostname-rewrite example](https://github.com/vercel/examples/tree/main/edge-functions/hostname-rewrites).

**Reply by steven-tey** [reply]
Thanks a lot for sharing this, @styxlab! Just to add on to this – we're also rolling out more complex boilerplates (e-commerce storefront builders, website builders) + supplementary code snippets (static tweets, domains API, OG image generator) in an [early access program](https://eap.vercel.sh/) – feel free to join if you're interested (click "sign up" on the Platforms on Vercel program).

**Reply by styxlab** [reply]
@steven-tey: Thanks for the link. I signed up, but did not see a way to access the content. Will I get notified?

**Reply by steven-tey** [reply]
@styxlab thanks for signing up! Yes, you will receive an invitation to join the Github repo when it's ready! 😇

**Reply by murbanowicz** [reply]
@steven-tey do I get it right that at this point signing up has no effect and we just have to wait for some communication?

**Reply by steven-tey** [reply]
@murbanowicz yup! You won't get an email when you sign up now but you will get an email invitation to join the private repo when we add you! :)

**Reply by murbanowicz** [reply]
Thanks! Can't wait!

**Reply by stevemarksd** [reply]
Would you have an example? I can't really see anything there that can also rewrite client side router. @styxlab

**Reply by styxlab** [reply]
> Would you have an example? I can't really see anything there that can also rewrite client side router. @styxlab 

Did you see https://github.com/vercel/platforms ?

**Comment by steven-tey** [comment, marked as answer]
Quick update on this – we recently launched the Platforms Starter Kit that's a comprehensive template for building multi-tenant applications with built-in custom domains support (+ unlimited custom domains 😉): https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Reply by steven-tey** [reply]
CC: @styxlab @murbanowicz @threehams @acorcutt @Skarian @iNishant @surjithctly @wmattei @ramiel @tomsoderlund @petrbela

**Reply by piyushhanchate** [reply]
Is there a similar starter kit for the pages router? I will not be be able to use the app router since I extensively use Material UI and NextJS app router doesn't support it yet
